casanovo/casanovo.py

```diff
@@ -145,5 +166,112 @@ def main(
+        "+43.006": "[UNIMOD, UNIMOD:5, Carbamyl, ]",
+        "-17.027": "[UNIMOD, UNIMOD:385, Ammonia-loss, ]",
+    }
+    residues = collections.defaultdict(set)
+    for aa, mass in config["residues"].items():
+        aa_mod = re.match(r"([A-Z]?)([+-]?(?:[0-9]*[.])?[0-9]+)", aa)
+        if aa_mod is None:
+            residues[aa].add(None)
+        else:
+            residues[aa_mod[1]].add(aa_mod[2])
+    fixed_mods, variable_mods = [], []
+    for aa, mods in residues.items():
+        if len(mods) > 1:
+            for mod in mods:
+                if mod is not None:
+                    variable_mods.append((aa, mod))
+        elif None not in mods:
+            fixed_mods.append((aa, mods.pop()))
+
+    # Write the mzTab output file header.
+    metadata = [
+        ("mzTab-version", "1.0.0"),
+        ("mzTab-mode", "Summary"),
+        ("mzTab-type", "Identification"),
+        (
+            "description",
+            f"Casanovo identification file "
+            f"{os.path.splitext(os.path.basename(filename_out))[0]}",
+        ),
+        (
+            "ms_run[1]-location",
+            pathlib.Path(os.path.abspath(filename_in)).as_uri(),
+        ),
+        (
+            "psm_search_engine_score[1]",
+            "[MS, MS:1001143, search engine specific score for PSMs, ]",
+        ),
+        ("software[1]", f"[MS, MS:1003281, Casanovo, {__version__}]"),
+    ]
+    if len(fixed_mods) == 0:
+        metadata.append(
+            (
+                "fixed_mod[1]",
+                "[MS, MS:1002453, No fixed modifications searched, ]",
+            )
+        )
+    else:
+        for i, (aa, mod) in enumerate(fixed_mods, 1):
+            metadata.append(
+                (
+                    f"fixed_mod[{i}]",
+                    known_mods.get(mod, f"[CHEMMOD, CHEMMOD:{mod}, , ]"),
+                )
+            )
+            metadata.append((f"fixed_mod[{i}]-site", aa if aa else "N-term"))
+    if len(variable_mods) == 0:
+        metadata.append(
+            (
+                "variable_mod[1]",
+                "[MS, MS:1002454, No variable modifications searched,]",
+            )
+        )
+    else:
+        for i, (aa, mod) in enumerate(variable_mods, 1):
+            metadata.append(
+                (
+                    f"variable_mod[{i}]",
+                    known_mods.get(mod, f"[CHEMMOD, CHEMMOD:{mod}, , ]"),
+                )
+            )
+            metadata.append(
+                (f"variable_mod[{i}]-site", aa if aa else "N-term")
+            )
+    for i, (key, value) in enumerate(kwargs.items(), 1):
+        metadata.append((f"software[1]-setting[{i}]", f"{key} = {value}"))
+    for i, (key, value) in enumerate(config.items(), len(kwargs) + 1):
+        if key not in ("residues",):
+            metadata.append((f"software[1]-setting[{i}]", f"{key} = {value}"))
+    with open(filename_out, "w") as f_out:
+        writer = csv.writer(f_out, delimiter="\t")
```

Comment:
**suggestion:** I think I agree with @wsnoble that we should use OS-specific line endings. What do you think?

It should be easy enough:
```python
writer = csv.writer(f_out, delimiter="\t", lineterminator=os.linesep())
```

Reply:
**praise:** Excellent solution.